File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // Figure out how many nodes are in the linkedlist
        int list_length = 0;

        ListNode counter = head;

        while(counter != null){
            // Increment list_length
            list_length++;
            // Move on to the next node
            counter = counter.next;
        }

        // if number of nodes is equal to n
        if(list_length == n){
            // then set new head to be the next node after head
            head = head.next;
        } else {
            // create a variable called p which is the length of the linked list
            int p = list_length;

            // create curr and prev pointers
            ListNode prev = new ListNode(0, head);
            ListNode curr = head;

            // go through the linkedlist from the head
            while(curr != null){
                // check if n is equal to p
                if(n == p){
                    // found the node to remove
                    prev.next = curr.next;
                    break;
                } 

                // decrease counter p
                p--;
                prev = prev.next;
                curr = curr.next;
            }
        }

        return head;
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        oldToCopy = { None: None}

        curr = head
        # creating the copied nodes
        while curr != None:
            copy = Node(curr.val)
            oldToCopy[curr] = copy
            curr = curr.next
        
        curr = head
        # connecting the copied nodes
        while curr != None:
            # get the copied nodes
            copy = oldToCopy[curr]
            copy.next = oldToCopy[curr.next]
            copy.random = oldToCopy[curr.random]
            curr = curr.next
        
        return oldToCopy[head]

        